lib/honeybadger/context_manager.rb

```diff
@@ -17,14 +17,28 @@ def clear!
     # Internal accessors
 
     def set_context(hash)
+      raise TypeError, "no implicit conversion of #{hash.class.name} into Hash" unless hash.respond_to?(:to_hash)
+
       @mutex.synchronize do
-        @context ||= {}
-        @context.update(hash)
+        @context ||= []
+        @context.push(hash)
+      end
+
+      return nil unless block_given?
+
+      begin
+        yield
+        nil
+      ensure
+        @mutex.synchronize { @context.delete(hash) }
```

Comment:
The call to `delete` here does worry me. It doesn't just remove the most recently added matching hash. It removes all matching hashes. 

Here's an IRB session showing what I mean:

```
irb(main):007:0> h1 = {a: 1, b:2}
=> {:a=>1, :b=>2}
irb(main):008:0> h2 = {a: 1, b:2}
=> {:a=>1, :b=>2}
irb(main):009:0> a = []
=> []
irb(main):010:0> a.push(h1)
=> [{:a=>1, :b=>2}]
irb(main):011:0> a.push(h2)
=> [{:a=>1, :b=>2}, {:a=>1, :b=>2}]
irb(main):012:0> a.delete(h1)
=> {:a=>1, :b=>2}
irb(main):013:0> a
=> []
```